### ISSUE TITLE
Elasticsearch: No option for 'nested' type bucket aggregation to deal with nested fields

### ISSUE THREAD

**Issue by arcolife** [grafana/grafana#4693]
Currently it's not possible to do bucket aggregation for nested type docs, even if PR #4527 would add nested type autocomplete feature. The query editor mode was removed, and even if it were there, grafana currently doesn't know how to deal with nested bucketed aggregation results. Would be nice to have this sorted out too.
Cheers!
## 
- What grafana version are you using?
  3.0.0 (latest from master branch)
- What datasource are you using?
  elasticsearch
- What OS are you running grafana on?
  fedora 23

**Comment by arcolife** [comment]
PR https://github.com/grafana/grafana/pull/4694 takes care of this. 

A working example:
![screenshot from 2016-04-19 20-53-19](https://cloud.githubusercontent.com/assets/920859/14645554/c00a85d0-0673-11e6-9480-5d98b948d2de.png)

WIP: Have to make changes to add autocomplete to the "nested query" field and add unit tests for all the additions.. I've also included changes from PR #4527 in this one. Maybe we could use github's new PR squash to merge them.

**Comment by vaibhavinbayarea** [comment]
Thanks for working on this feature request. Ability to work with nested objects will make Grafana a lot more useful than Kibana. 

I was looking at the implementation and it works great! One enhancement I would suggest (we can track it as a different request / issue once this is merged in), is ability to plot multiple series based on nested property rather than just one series with the filter option.

For example, 

My field value for data in elasticsearch source looks like:

```
ipmi_sensors.fan = [ {
  "name": "FAN1",
  "id": 339,
  "units": "RPM",
  "reading": 4125,
  "event": "'OK'"
},
{
  "name": "FAN2",
  "id": 406,
  "units": "RPM",
  "reading": null,
  "event": null
},
{
  "name": "FAN3",
  "id": 473,
  "units": "RPM",
  "reading": 4200,
  "event": "'OK'"
},
{
  "name": "FAN4",
  "id": 540,
  "units": "RPM",
  "reading": 4050,
  "event": "'OK'"
},
{
  "name": "FAN5",
  "id": 607,
  "units": "RPM",
  "reading": 6000,
  "event": "'OK'"
},
{
  "name": "FAN6",
  "id": 674,
  "units": "RPM",
  "reading": null,
  "event": null
},
{
  "name": "FANA",
  "id": 741,
  "units": "RPM",
  "reading": null,
  "event": null
},
{
  "name": "FANB",
  "id": 808,
  "units": "RPM",
  "reading": null,
  "event": null
}]
```
Its important when querying, max of FAN1 reading is 4125 and not 6000 (which is max across the document).

With your patch, I can create 6 queries, 1 for each name and filter based on name. This works great, but we should provide option of automatic population of series based on the field terms "name". Ideally support both filter based OR term based.

To get unique streams per term name, my elasticsearch query looks like:
```
{
   "size":0,
   "query":{
      "filtered":{
         "query":{
            "query_string":{
               "analyze_wildcard":true,
               "query":"customer_name:XYZ AND server_uuid:XXXXX"
            }
         },
         "filter":{
            "bool":{
               "must":[
                  {
                     "range":{
                        "@timestamp":{
                           "gte":"1469473732334",
                           "lte":"1470078532334",
                           "format":"epoch_millis"
                        }
                     }
                  }
               ]
            }
         }
      }
   },
   "aggs":{
      "7":{
         "date_histogram":{
            "interval":"10m",
            "field":"@timestamp",
            "min_doc_count":0,
            "extended_bounds":{
               "min":"1469473732334",
               "max":"1470078532334"
            },
            "format":"epoch_millis"
         },
         "aggs":{
            "6":{
               "nested":{
                  "path":"ipmi_sensors.fan"
               },
               "aggs":{
                  "nested_aggs":{
                     "terms":{
                        "field":"ipmi_sensors.fan.name"
                     },
                     "aggs":{
                        "1":{
                           "max":{
                              "field":"ipmi_sensors.fan.reading"
                           }
                        }
                     }
                  }
               }
            }
         }
      }
   }
}
```

My patch / changes made to use Terms instead of Filter is:
```
To query and display the results correctly, I made the following changes:
diff --git a/public/app/plugins/datasource/elasticsearch/elastic_response.js b/public/app/plugins/datasource/elasticsearch/elastic_response.js
index 66f8dad..d420b1a 100644
--- a/public/app/plugins/datasource/elasticsearch/elastic_response.js
+++ b/public/app/plugins/datasource/elasticsearch/elastic_response.js
@@ -98,29 +98,40 @@ function (_, queryDef) {
   };
 
   ElasticResponse.prototype.processNestedAggregationDocs = function(esAgg, aggDef, target, seriesList, props) {
-    var metric, y, i, newSeries, bucket, value;
+    var metric, y, i, j, newSeries, outer_bucket, bucket, value;
 
     for (y = 0; y < target.metrics.length; y++) {
       metric = target.metrics[y];
       if (metric.hide) {
         continue;
       }
-
+      var seriesHash = {};
       newSeries = { datapoints: [], metric: metric.type, field: metric.field, props: props};
       for (i = 0; i < esAgg.buckets.length; i++) {
-        bucket = esAgg.buckets[i][aggDef.id]['nested_aggs'];
-        if (bucket !== undefined) {
-          value = bucket[metric.id];
-          if (value !== undefined) {
-            if (value.normalized_value) {
-              newSeries.datapoints.push([value.normalized_value, esAgg.buckets[i].key]);
-            } else {
-              newSeries.datapoints.push([value.value, esAgg.buckets[i].key]);
+        outer_bucket = esAgg.buckets[i][aggDef.id]['nested_aggs'];
+        for (j =0; j< outer_bucket.buckets.length; j++) {
+          bucket = outer_bucket.buckets[j];
+          if (bucket !== undefined) {
+            value = bucket[metric.id];
+            // add key if it doesnt exist
+            if (!(bucket.key in seriesHash)) {
+              var pr = {};
+              pr[aggDef.settings.nested.term] = bucket.key;
+              seriesHash[bucket.key] = { datapoints: [], metric: metric.type, field: metric.field, props: pr};
+            }
+            if (value !== undefined) {
+              if (value.normalized_value) {
+                seriesHash[bucket.key].datapoints.push([value.normalized_value, esAgg.buckets[i].key]);
+              } else {
+                seriesHash[bucket.key].datapoints.push([value.value, esAgg.buckets[i].key]);
+              }
             }
           }
         }
       }
-      seriesList.push(newSeries);
+      for (var seriesName in seriesHash) {
+        seriesList.push(seriesHash[seriesName]);
+      }
     }
   };
 
diff --git a/public/app/plugins/datasource/elasticsearch/query_builder.js b/public/app/plugins/datasource/elasticsearch/query_builder.js
index 7a64b0a..4c4e2fe 100644
--- a/public/app/plugins/datasource/elasticsearch/query_builder.js
+++ b/public/app/plugins/datasource/elasticsearch/query_builder.js
@@ -62,12 +62,13 @@ function (queryDef) {
 
     queryNode.aggs = {
       nested_aggs: {
-        filter: {},
+        terms: {},
         aggs: {}
       }
     };
-    queryNode.aggs.nested_aggs.filter.term = {};
-    queryNode.aggs.nested_aggs.filter.term[settings.nested.term] = settings.nested.query;
+    queryNode.aggs.nested_aggs.terms = { "field": settings.nested.term };
+    //queryNode.aggs.nested_aggs.filter.term = {};
+    //queryNode.aggs.nested_aggs.filter.term[settings.nested.term] = settings.nested.query;
 
     queryNode.aggs.nested_aggs.aggs = {};
     for (y = 0; y < target.metrics.length; y++) {

```

The results are:
![screen shot 2016-08-01 at 12 18 19 pm](https://cloud.githubusercontent.com/assets/4401392/17305942/190557fe-57e2-11e6-8bbf-5cf1eab23298.png)

I'm presently using my patch to display graphs of all my nested objects (ntp, ipmi, smart data) and seems to be working well.

Let me know what you guys think! Probably will be a different feature request unless you can pull it the changes and add support for TERM based series :)

**Comment by arcolife** [comment]
@vaibhavinbayarea excellent! and apologies I couldn't go through this any sooner. Although I'll review this and reply soon, but PR #4694 which fixes the above issue, is currently still pending review from the maintainers. If you wanna submit this new addition to my branch (arcolife/nested_agg_query) as a PR (since you suggested changes to my code above), I'd be glad to include that, or else I could just include the additions as you've mentioned above.

**Comment by vaibhavinbayarea** [comment]
@arcolife sorry for the late response.. I was away for a while. I tried combining my patch with yours but had a few issues. If I'm correct, your implementation does filter based aggregation. My patch above does term based aggregation. We can't do both at the same time. 

Elasticsearch responds with:

> 
reason": "Found two aggregation type definitions in [nested_aggs]: [terms] and [filter]" 

In terms based aggregation, there is an "include" field, which filters based on regex (https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-terms-aggregation.html#_filtering_values_2) but its not same as ES query language.

So for both to exist side by side, one alternative I see is to add a checkbox/dropdown which specifies the type of aggregation being requested ("filter"/"terms") and based on the response, get a query or a regex. 

What do you think? 

Meanwhile, I'll keep looking if there is a way to filter query while doing terms aggregation or vice versa (hopefully, I'm not missing something obvious).

**Comment by arcolife** [comment]
@torkelo could you reopen this? ref: [your comment](https://github.com/grafana/grafana/pull/4694#issuecomment-282676410)

**Comment by arcolife** [comment]
updates at https://github.com/grafana/grafana/pull/7863#issuecomment-354004166

**Comment by kaiwalyapowle** [comment]
I downloaded the committed source code from https://github.com/grafana/grafana/tree/8b350c49acaf55d31f27f3285013ab0340d33047.

Can someone please tell me how to run the grafana service from the code downloaded from the above on windows ?

**Comment by FrankOrtmann** [comment]
Really need this feature, sad it is not getting any love.

**Comment by vivkv07** [comment]
Will be glad if either Grafana or Kibana comes up with this feature. This will make our lives easier to track performances. Any other opensource platforms which support this feature ?

**Comment by ameed93** [comment]
+1

**Comment by tammytankian** [comment]
+1 please, oh please

**Comment by ivan046** [comment]
+1

**Comment by MoeSaidi** [comment]
+1

**Comment by M0rdecay** [comment]
+1

**Comment by deepaksuseelan** [comment]
+1

**Comment by valleedelisle** [comment]
What else do we need to complete this feature? I see a bunch of stuff got merged but we still don't have this feature which is, imho, very important.

**Comment by arcolife** [comment]
It's been 4+ years now since I implemented this and related PR work. The reason I've gotten thus far is "feature creep". https://github.com/grafana/grafana/pull/7863#issuecomment-455481912

**Comment by mikeshimkus** [comment]
+1

**Comment by martini** [comment]
+1

**Comment by venturieffect** [comment]
+1

**Comment by martini** [comment]
+1

**Comment by MrGeneration** [comment]
This would actually really help our use cases a lot. 🙈

**Comment by mrarmyant** [comment]
Where is this?

**Comment by ashishkaransingh** [comment]
We need this.

**Comment by unreal-altran** [comment]
is there a delivery plan that provides this functionality in grafana 7.x?

**Comment by dwtien** [comment]
+1

**Comment by jehutywong** [comment]
+1

**Comment by nguedon** [comment]
+1

**Comment by nguedon** [comment]
+1 in grafana 8 ?

**Comment by Amooti** [comment]
+1

**Comment by wols** [comment]
After a test migration of my data flow to support [Elastic Common Schema (ECS)](https://www.elastic.co/guide/en/ecs/1.11/ecs-reference.html) the horror came:
- all my fields are nested now
- Grafana (8.0.6 here) doesn't support this :-1: 

**With the spread of ECS, the usability of Elasticsearch datasource with Grafana will continue to decline.** 

The [Elasticsearch Writer](https://icinga.com/docs/icinga-2/latest/doc/14-features/#elasticsearch-writer) from Icinga2 also provides such nested data.
See also https://community.grafana.com/t/elasticsearch-template-variable-with-nested-fields/9709
I expect a large community that would welcome the 'nested' feature (since 2018) in Grafana.

The Table panel shows the nested values correctly `Metric: Raw Data`.

![grafana_table_raw](https://user-images.githubusercontent.com/764167/129635251-6df02737-8cbd-42c8-8a96-9135cb760054.png)

Interestingly, in Time series panel when grouping, the nested field is accessed.

![grafana_timeseries_group](https://user-images.githubusercontent.com/764167/129636069-229e3014-b1a7-4a56-85ed-781f214bad6a.png)

**Comment by Niek** [comment]
+1

**Comment by davidcorbin** [comment]
+1

**Comment by madhukar93** [comment]
+1

**Comment by JustKiddingCode** [comment]
So no that this Feature Request has priority/customer-support, may be a good time to open/work on/ merge https://github.com/grafana/grafana/pull/7863
It seems to me the most recent PR for this issue.

**Comment by StefanVahlgren** [comment]
+1. How can this not be solved yet? If I'm not understanding the issue description, solving this would help many people. I need to be able to use values from the tags array in the panel query in the "group by" part. 
"process": {
"serviceName": "my-service",
"tags": [
{
"key": "hostname",
"type": "string",
"value": "my-hostname"
},.....

**Comment by SudoNova** [comment]
Dear devs
as you know Elasticsearch advises to use nested fields instead of normalizing data, as there is no SQL-like joins available. Therefore accessing nested fields is a must. I kindly ask you to devote more attention to this issue.
Best regards

**Comment by AndreBaumeier** [comment]
and another bump. If you're using elasticsearch as it is intended, denormalized, document oriented you're not really able to query your data in grafana.

**Comment by adnanel** [comment]
any update on this? there are so many issues talking about this i don't even know where to ask.

**Comment by IamShubhamGupto** [comment]
@Elfo404 Please implement this feature / provide a workaround for displaying graphs on nested data

thanks

**Comment by ethan-gallant** [comment]
+1

**Comment by mattjtodd** [comment]
+1

**Comment by ethan-gallant** [comment]
Anyone who is looking for this feature still, please leave a reaction on my latest PR. https://github.com/grafana/grafana/pull/47233

**Comment by RowanWally** [comment]
Hello,
Any update on this?

**Comment by ethan-gallant** [comment]
@RowanWally I've updated my PR here https://github.com/grafana/grafana/pull/47233 am waiting to find out what else needs to happen before it can merge.

**Comment by ethan-gallant** [comment]
@gabor a month too late but this can be closed since https://github.com/grafana/grafana/pull/47233 -> https://github.com/grafana/grafana/pull/62301 was merged into main
